src/main/kotlin/Task.kt

```diff
@@ -0,0 +1,98 @@
+
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.serializer
+import org.slf4j.LoggerFactory
+import java.util.*
+import kotlin.time.Duration
+
+
+class Task<T: Any> @PublishedApi internal constructor(
+    private val inputSerializer: KSerializer<T>,
+    val name: String,
+    val retry: IRetryPolicy,
+    val handler: TaskHandlerWithContext<T>
+) {
+    private var logger = LoggerFactory.getLogger(this::class.java)
+
+    companion object {
+        inline fun <reified T : Any> create(
+            name: String, retry: IRetryPolicy = DefaultRetryPolicy(), noinline handler: TaskHandlerWithContext<T>
+        ) = Task(serializer(), name, retry, handler)
+
+        inline fun <reified T : Any> create(
+            name: String, retry: IRetryPolicy = DefaultRetryPolicy(), noinline handler: TaskHandler<T>
+        ) = create(name, retry, handler.toWithContext())
+    }
+
+    fun callLater(input: T, params: CallParams = CallParams(), manager: TaskManager = TaskManager.getDefaultInstance()) {
+        val inputStr = Json.encodeToString(inputSerializer, input)
+        logger.debug("Call task $name later with input $inputStr")
+        manager.enqueueTaskCall(this, inputStr, params)
+
+    }
+
+    fun execute(inputStr: String, params: CallParams, manager: TaskManager) {
+        logger.debug("Execute task $name with input $inputStr")
+        // TODO: what to do with deserialization errors?
```

Comment:
Just a thought: task and brokers can have a version (defined in code and fixed at compile time)
If some broker finds a message he cannot read, he updates message's RetryPolicy to something like (waitForVersion = currentVersion + 1)

Reply:
My vision is different. Every time a developer needs to change task's input or significantly change the behaviour they need to change the task's name. So we never have mixed data In a queue.

> , he updates message's RetryPolicy to something like (waitForVersion = currentVersion + 1)
`RetryPolice` is not serialised into a message. It exists only in the code, so it can't be changed far a single task call